client/src/shared/state/Inventory.js

```diff
@@ -104,16 +107,155 @@ class Inventory {
         if (!ItemTypes[itemConfig.typeCode].hasUseEffect) return;
 
         this.hotbar.push(itemConfig);
-
+        this.keyToSlotIndex[this.hotbar.length - 1] = itemConfig.slotIndex;
         PubSub.publish(HOTBAR_ITEM);
         PubSub.publish(MODIFY_INVENTORY_ITEM);
+        if (this.saveSession === true) {
+            this.saveHotbar();
+        }
+    }
+
+    resetMapping() {
+        Utils.warning("This is a bug found or a localStorage mismatch");
+        this.keyToSlotIndex = {};
+        this.saveHotbar();
+        this.hotbar = [];
+        this.defaultHotBar();
     }
 
-    removeFromHotbar(itemConfig) {
+    updateSlotIndex(slotIndex) {
+        Object.keys(this.keyToSlotIndex).forEach((key, index) => {
+            if (this.keyToSlotIndex[key] >= slotIndex) {
+                const updateValue = this.keyToSlotIndex[key] - 1;
+                if (updateValue >= 0) {
+                    this.keyToSlotIndex[key] = updateValue;
+                    if (this.saveSession === true) {
+                        this.saveHotbar();
+                    }
+                }
+            }
+        });
+    }
+
+    updateMapHotBarKeys(evictKey) {
+        Object.keys(this.keyToSlotIndex).forEach((key, index) => {
+            if (key >= evictKey) {
+                const updateKey = key - 1;
+                if (updateKey >= 0) {
+                    this.keyToSlotIndex[updateKey] = this.keyToSlotIndex[key];
+                    delete this.keyToSlotIndex[key];
+                }
+            }
+        });
+    }
+
+    updateMapSlotIndex(evictValue) {
+        Object.keys(this.keyToSlotIndex).forEach((key, index) => {
+            if (this.keyToSlotIndex[key] >= evictValue) {
+                const newValue = this.keyToSlotIndex[key] - 1;
+                if (newValue >= 0) {
+                    this.keyToSlotIndex[key] = newValue;
+                }
+            }
+        });
+    }
+
+    removeFromHotbar(itemConfig, removedInventory = false) {
```

Comment:
booleans in function parameters are a code smell: https://understandlegacycode.com/blog/what-is-wrong-with-boolean-parameters/

You might want to break this function apart into smaller functions that perform the necessary isolated behaviours.

What does it mean when you call this function with `removedInventory=true`, does it mean the player dropped an item from their inventory and so consequentially the item must also be removed from the hotbar?

Reply:
Yes, pretty much, removed from inventory so, update both the key and value, instead of updating just the key, I will see if time permits, to break it up.

Reply:
Yeah, from my through playtesting I have not encountered an issue, for now I am gonna leave it.